Performance Loss from OpenCV 4.5.5 to 4.7.0 using CUDA backend

### System Information

OpenCV versions tested: 4.5.5, 4.7.0
Operating System / Platform: Ubuntu 18.04
Device: NVIDIA Jetson TX2 DevKit
CUDA version: 10.2
CUDNN version: 8.2.1

### Detailed description

Hi,

I was using OpenCV 4.5.5, backend CUDA on a NVIDIA Jetson TX2 Devkit with the specs defined above. A couple of days I decided to update to OpenCV 4.7.0 to check if I had some boost in performance for the models I'm currently using. However what I did saw was a performance loss (in terms of execution time) for the majority of the models. Do you know what is the reason for this loss of performance?

This is the execution times obtained for both versions of OpenCV:

<html>
<body>
<!--StartFragment--><h2>Test 1</h2><ul><li>Device: <strong>TX2 DevKit</strong></li><li><strong>CUDA </strong>version: <strong>10.2</strong></li><li><strong>CUDNN</strong> version: <strong>8.2.1</strong></li><li><strong>OpenCV </strong>version:<strong> 4.7.0</strong></li></ul><p><br /></p>

Version | Model 1 | Model 2 | Model 3 | Model 4
-- | -- | -- | -- | --
Input Size | (112, 112) | (112, 112) | (112, 112) | (112, 112)
Model Architecture | Resnet100 | MobileFaceNet | Resnet100 | Resnet18
Jetson CPU | 702 | 20.5 | 699 | 167
Jetson GPU | **91.7** | **10.5** | **91.6** | **52.2**

<!--EndFragment-->
</body>
</html>

<html>
<body>
<!--StartFragment--><h2>Test 2</h2><ul><li>Device: <strong>TX2 DevKit</strong></li><li><strong>CUDA </strong>version: <strong>10.2</strong></li><li><strong>CUDNN</strong> version: <strong>8.2.1</strong></li><li><strong>OpenCV </strong>version:<strong> 4.5.5</strong></li></ul><p><br /></p>

Version | Model 1 | Model 2 | Model 3 | Model 4
-- | -- | -- | -- | --
Input Size | (112, 112) | (112, 112) | (112, 112) | (112, 112)
Model Architecture | Resnet100 | MobileFaceNet | Resnet100 | Resnet18
Jetson CPU | 1088 | 23.1 | 1096 | 257
Jetson GPU | **60.9** | **5.34** | **60.7** | **19.9**

<!--EndFragment-->
</body>
</html>

**Note:** Both tests were built with the same OpenCV flags and requirements, the only thing that changed was the version of both opencv and opencv_contrib. Moreover, all the execution times presented in those tables are in ms.

### Steps to reproduce

You can use this piece of code to reproduce this issue/loss of performance:

```
#include <thread>
#include <fstream>

#include <opencv2/imgproc.hpp>
#include <opencv2/dnn.hpp>
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>

int main(int argc, char** argv)
{
    auto imageToTest = argv[1];
    auto modelToTest = argv[2];
    int modelInputWidth = atoi(argv[3]);
    int modelInputHeight = atoi(argv[4]);

    cv::Size currSize = cv::Size(modelInputWidth, modelInputHeight);
    std::string modelToTestOnnx = modelToTest;
    std::string imagefilename = imageToTest;
    unsigned int num_inferences = 100;

    cv::dnn::Net net = cv::dnn::readNetFromONNX(modelToTestOnnx);

    net.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
    net.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA);

    cv::Mat img = cv::imread(imagefilename, cv::IMREAD_ANYCOLOR);
    cv::Mat resized;
    cv::resize(img, resized, currSize);
	
	std::vector<cv::Mat> imgBatch = { resized };
    bool swaprbchannels = false;
    cv::Mat blob = cv::dnn::blobFromImages(imgBatch, 1.0f / 255.0f, cv::Size(), cv::Scalar(), swaprbchannels, false, CV_32F);

    net.setInput(blob);

    std::vector<cv::String> unconnectedOutLayerNames = net.getUnconnectedOutLayersNames();

    std::vector<cv::Mat> outputs;
    outputs.clear();

    auto timeLoadModelPlusInference1 = std::chrono::high_resolution_clock::now();

    net.forward(outputs, unconnectedOutLayerNames);

    auto timeLoadModelPlusInference2 = std::chrono::high_resolution_clock::now();

    std::chrono::duration<double, std::milli> ms_doubleTimeLoadModelPlusInference = timeLoadModelPlusInference2 - timeLoadModelPlusInference1;

    std::cout << "Execution time (load model + inference): " << ms_doubleTimeLoadModelPlusInference.count() << std::endl; // in ms

    auto time1 = std::chrono::high_resolution_clock::now();

    try {
        for (size_t i = 0; i < num_inferences; i++)
		net.forward(outputs, unconnectedOutLayerNames);
    }
    catch (std::exception& ex)
    {
        std::cout << ex.what() << std::endl;
    }

    auto time2 = std::chrono::high_resolution_clock::now();

    std::chrono::duration<double, std::milli> ms_double = time2 - time1;
    std::cout << "Execution time inference only: " << ms_double.count() / num_inferences << std::endl; // in ms

    std::cout << "Outputs Size: " << outputs[0].size[0] << "x" << outputs[0].size[1] << std::endl;
    std::cout << "Outputs value: " << outputs[0] << std::endl;
}

```
 

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

So I just built with OpenCV 4.x (**4.7.0-dev**) and this is the updated table.

<html>
<body>
<!--StartFragment--><h2>2.5. Version 5</h2><ul><li>Device: <strong>TX2 DevKit</strong></li><li><strong>CUDA </strong>version: <strong>10.2</strong></li><li><strong>CUDNN</strong> version: <strong>8.2.1</strong></li><li><strong>OpenCV </strong>version:<strong> 4.X (4.7.0-dev)</strong></li></ul><p><br /></p>

Version | Model 1 | Model 2 | Model 3 | Model 4
-- | -- | -- | -- | --
Input Size | (112, 112) | (112, 112) | (112, 112) | (112, 112)
Model Architecture | Resnet100 | MobileFaceNet | Resnet100 | Resnet18
Jetson CPU | ERROR | ERROR | ERROR | 166
Jetson GPU | ERROR | ERROR | ERROR | 34.7

<p><strong>ERROR:&nbsp;</strong></p><p>terminate called after throwing an instance of 'cv::Exception'<br />what(): OpenCV(4.7.0-dev) /home/vbuser/opencv/modules/core/src/matrix.cpp:1177: error: (-211:One of the arguments' values is out of range) Bad new number of rows in function 'reshape'</p><!--EndFragment-->
</body>
</html>

So, Model 1, 2, and 3 are crashing now (on branch 4.7.0-dev: 4.x) and although Model 4 has now a better performance than release 4.7.0 (from 52.2 to 34.7), his performance (in terms of execution time) is not better than release 4.5.5 (from 19.9 to 34.7).

--- a/modules/dnn/perf/perf_layer.cpp
+++ b/modules/dnn/perf/perf_layer.cpp
@@ -66,8 +66,13 @@ struct Layer_NaryEltwise : public TestBaseWithParam<tuple<Backend, Target> >
 
         if (!isRef && backendId == DNN_BACKEND_CUDA)
         {
-            if (a_shape != b_shape)
-                throw SkipTestException("The test is skipped because inputs with different shapes are not supported.");
+            if (a_shape.size() != b_shape.size())
+                throw SkipTestException("The test is skipped because inputs with different shape size are not supported.");
+
+            for(int i = 0; i < a_shape.size(); i++)
+                if (a_shape[i] != b_shape[i] && a_shape[i] != 1 && b_shape[i] != 1)
+                    throw SkipTestException("The test is skipped because inputs are not supported.");
+
             if (nary_eltwise_cuda_deny_ops.find(op) != nary_eltwise_cuda_deny_ops.end())
                 throw SkipTestException("The operator '" + op + "' is skipped because is not support with cuda currently.");
         }
@@ -213,6 +218,11 @@ PERF_TEST_P_(Layer_NaryEltwise, NCHW_C_sum)
 PERF_TEST_P_(Layer_NaryEltwise, NHWC_C)
 {
     test_layer({N, H, W, C}, {1, C}, "sum");
+}
+
+PERF_TEST_P_(Layer_NaryEltwise, NHWC_H)
+{
+    test_layer({N, H, W, C}, {1, H, 1, 1}, "sum");
 }
 
 PERF_TEST_P_(Layer_Slice, YOLOv4_tiny_1)

--- a/modules/dnn/src/layers/nary_eltwise_layers.cpp
+++ b/modules/dnn/src/layers/nary_eltwise_layers.cpp
@@ -673,12 +673,17 @@ public:
     {
         auto context = reinterpret_cast<csl::CSLContext*>(context_);
 
-        auto input_wrapper = inputs[0].dynamicCast<CUDABackendWrapper>();
+        auto input_0_shape = inputs[0].dynamicCast<CUDABackendWrapper>()->getShape();
         for (int i = 1; i < inputs.size(); i++)
         {
-            auto from_wrapper = inputs[i].dynamicCast<CUDABackendWrapper>();
-            if (input_wrapper->getShape() != from_wrapper->getShape())
+            auto input_i_shape = inputs[i].dynamicCast<CUDABackendWrapper>()->getShape();
+            if (input_0_shape.size() != input_i_shape.size())
                 return Ptr<BackendNode>();
+            // check if the shape can be supported by `eltwise_ops.cu`, or return the default BackendNode
+            for (int j = 0; j < input_0_shape.size(); j++)
+                if (input_0_shape[j] != input_i_shape[j] &&
+                    input_0_shape[j] != 1 && input_i_shape[j] != 1)
+                    return Ptr<BackendNode>();
         }
 
         cuda4dnn::EltwiseOpType op_ = cuda4dnn::EltwiseOpType::SUM;

